man/man1/runlpvi.1
.TH RUNLPVI 1NEMO "2 January 2020"
.SH NAME
runlpvi \- NEMO front end for the LP-VIcode
.SH SYNOPSIS
\fBrunlpvi\fP [parameter=value]
.SH DESCRIPTION
\fBrunlpvi\fP is a NEMO front-end to the LP-VIcode, which computes
a set of chaos indicators for (galactic) orbits. The main documentation can be
found online in their manual
http://www.fcaglp.unlp.edu.ar/LP-VIcode/manual-LPVIcodeKaosv1.0.2.pdf . Here we
review the particulars of the NEMO user interface, see also EXAMPLES below for
some of the NEMO examples.
.PP
Compiling and Running the original \fILP-VIcode\fP requires 4 actions:
.PP
1) provide a "PAV" fortran coded file with use your Potential, Acceleration and
Variationals and include it in the code, and compile it. Note that in the
current implementation both the potential and its parameters are specified in
the executable, and unlike NEMO's potname/potpars interface cannot be changed.
.PP
2) write an initial conditions file (4,5,6 or 7 values on the line). We
only support 4 or 6 at the moment. (see \fBtstop=\fP below)
.PP
3) write a input parameter file. Needs to be 20 lines, including comments.
This needs to be named \fBLP-VIcode.in\fP.
.PP
4) write a parameter file in which NDIM is defined (the rest seems not important)
This needs to be named \fBLP-VIcode.par\fP.
.PP
This is sufficiently complex to make running many examples cumbersome, hence
the \fIrunlpvi\fP NEMO interface, which takes care of items 2-4.  Linking
the LP-VIcode is left as an exercize to the reader.
.SH PARAMETERS
.TP 20
\fBoutdir=\fP
Output run directory, which must not exist yet. (required) 
.TP
\fBposvel=\fP
Initial conditions (4 or 6 values). Multiple orbits can be given now, but without
individual times. The \fBndim=\fP is then required.  A hardcoded maximum is
1000. No default.
.TP
\fBprefix=\fP
Prefix for output files.  The name is not very important, as all files are
in a new directory.  The extensions are described in the manual,
but in brief are: \fI.ene\fP, \fI.orb\fP,  \fI.lyap\fP,
\fI.sali\fP, \fI.gali\fP, \fI.sd\fP, 
\fI.ssn\fP, \fI.rli\fP, \fI.megno\fP, and \fI.fli\fP.
[log]
.TP
\fBdt=\fP
Integration step [0.01]     
.TP
\fBtstop=\fP
Time of integration. In a future version multiple orbits
can have different integration times, which means they need
to be specified at the end of the \fBposvel=\fP sequence.
[1]    
.TP
\fBscreen=\fP
Screen output? [1]     
.TP
\fBorbit=\fP
Orbit output? [0]     
.TP
\fBci=\fP
CI indictors: LIs, SALI, GALIs, SD & SSNs, RLI & LImax, MEGNO & SElLCE, FLI & OFLI.
0=don't compute, 1=output for all t, 2=output only last value
.TP
\fBnstep=\fP
Nr. of steps between outputs (when orbit or indicators are = 1)
[1] 
.TP
\fBdev=\fP
Initial dev. vectors (0 = at random, 1 = random orthonormal, 2=fixed)
[1]
.TP
\fBsali=\fP
SALI & potential(t): normal saturation (=0) or restart
[1]
.TP
\fBndim=\fP
Not used yet, but might be needed if we allow multiple orbits and optional tstop's.
.TP
\fBexe=\fP
Name of the fortran executable [LP-VIcode]
.SH EXAMPLES
The \fIlogarithm2\fP example from the manual can be run as follows:
(2nd CI for the moment is 0, some error):
.nf

  runlpvi run1 posvel=0.2,0.34,0.5,0.1235 prefix=log dt=0.05 tstop=25000 orbit=1 ci=1,0,1::5 nstep=20
  tabplot run1/log.orb 2 3 -0.5 0.5 -0.5 0.5

.fi
which took about 35sec CPU. We follow with some pure NEMO orbits with the same initial conditions
and potential:
.nf

  mkorbit run1/orb.in 0.2 0.34 0 0.5 0.1235 0 potname=log potpars=0,2,0,0.7
  orbint run1/orb.in run1/orb.out  50000 0.05 ndiag=10000
  
.fi
and it becomes obvious the integrator in \fBorbint\fP cannot handle a 0 core radius for box orbits.
To show both method agree try this loop/tube orbit:
.nf

  mkorbit run2.in x=0.8 lz=1 e=1 potname=log potpars=0,0.5,0,0.7 
  orbint  run2.in - 2000 0.01 ndiag=200 | orbplot - yapp=2/xs 
  orblist run2.in
  #>> 0 0 0.8 0 0 0 1.56406 0  0 -1e+30 1e+30 7.35388e+188
  runlpvi run2 posvel=0.8,0,0,1.56406 prefix=log dt=0.01 tstop=20 orbit=1 ci=1::7 nstep=20
  tabplot run2/log.orb 2 3 -2 2 -2 2  line=1,1 yapp=1/xs

.fi
.SH SEE ALSO
orbdim(1NEMO)
.PP
.nf
http://lp-vicode.fcaglp.unlp.edu.ar/
http://www.fcaglp.unlp.edu.ar/LP-VIcode/manual-LPVIcodeKaosv1.0.2.pdf
.fi
.SH FILES
NEMO/usr/lpvi
.SH AUTHOR
Peter Teuben (runlpvi)
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
29-dec-2019	V0.1 Created	PJT
30-dec-2019	V0.3 posvel= now allows multiple orbits	PJT
.fi


usr/lpvi/runlpvi.c
/*
 *  runlpvi:   NEMO frontend for LP-VIcode 
 *	
 *  29-dec-2019  V0.1  initial draft                  PJT
 *                0.3  multiple orbits in posvel=
 *
 *  See also http://lp-vicode.fcaglp.unlp.edu.ar/
 *
 */

#include <stdinc.h>
#include <getparam.h>
#include <history.h>
#include <filefn.h>
#include <run.h>

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>

#include <vectmath.h>
#include <filestruct.h>


string defv[] = {
    "outdir=???\n        Output run directory (required)",
    "posvel=???\n        Initial conditions (4 or 6 values)",
    "prefix=log\n        Prefix for output files",
    "dt=0.01\n           Integration step",
    "tstop=1\n           Time of integration",
    "screen=1\n          Screen output?",
    "orbit=0\n           Orbit output?",
    "ci=0,0,0,0,0,0,0\n  CI indictors:  LIs, SALI, GALIs, SD & SSNs, RLI & LImax, MEGNO & SElLCE, FLI & OFLI",
    "nstep=1\n           Steps until Output",
    "dev=1\n             Initial dev. vectors (0 = at random, 1 = random orthonormal, 2=fixed)",
    "sali=0\n            SALI & potential(t): normal saturation (=0) or restart dev. vectors (=1)",

    "ndim=\n             Not used yet, but might be needed if posvel ambiguous",
    "exe=LP-VIcode\n     Name of the fortran executable",

    "VERSION=0.3\n       3-jan-2012 PJT",
    NULL,
};

string usage="NEMO front end for LP-VIcode";

#define MAXCI 7
#define MAXORB 1000

void nemo_main(void)
{
    int i, npv, ndim;
    string exefile = getparam("exe");
    string infile = "LP-VIcode.in";
    string parfile = "LP-VIcode.par";
    string datfile = "posvel.dat";
    string rundir = getparam("outdir");
    string fname;
    char dname[256], runcmd[256];
    real posvel[MAXORB*7];
    int nci, ci[MAXCI];
    stream ostr, datstr, histr;

    if (hasvalue("ndim"))
      ndim = getiparam("ndim");
    else
      ndim = -1;

    npv = nemoinpr(getparam("posvel"),posvel,MAXORB*7);
    if (npv==6 && ndim<0)
      ndim = 3;
    else if (npv==4 && ndim<0)
      ndim = 2;
    else if (ndim<0) 
      error("Need (multiple of) 4 or 6 values for posvel, or specifiy ndim=");

    nci = nemoinpi(getparam("ci"),ci,MAXCI);
    if (nci<0) error("Error ci=");
    for (i=nci; i<MAXCI; ++i) ci[i] = 0;

    run_mkdir(rundir);
    run_cd(rundir);

    // be nice and write a (NEMO) history file
    
    histr = stropen("history","w");
    put_history(histr);
    strclose(histr);


    // write the required .par file

    ostr = stropen(parfile,"w");
    fprintf(ostr,"* LP-VIcode.par : generated by NEMO's runlpvi\n");
    fprintf(ostr,"      INTEGER ndim\n");
    fprintf(ostr,"      PARAMETER (ndim=%d)\n",ndim);
    strclose(ostr);

    // write the initial conditions file
    // assume 1 orbit for now, no tstop per orbit

    ostr = stropen(datfile,"w");
    for (i=0; i<npv; ++i) {
      if (i>0 && i%(ndim*2)==0) fprintf(ostr,"\n");
      fprintf(ostr,"%f ",posvel[i]);
    }
    fprintf(ostr,"\n");
    strclose(ostr);

    // write the required .in file (comment lines are significant by the fact they need to exist)

    ostr = stropen(infile,"w");
    fprintf(ostr,"# LP-VIcode.in : generated by NEMO's runlpvi\n");
    fprintf(ostr,"# \n");
    fprintf(ostr,"%s\n",datfile);
    fprintf(ostr,"# \n");
    fprintf(ostr,"%s\n",getparam("prefix"));
    fprintf(ostr,"# \n");
    fprintf(ostr,"%g\n",getrparam("dt"));
    fprintf(ostr,"# \n");
    fprintf(ostr,"%g\n",getrparam("tstop"));
    fprintf(ostr,"# \n");
    fprintf(ostr,"%d %d\n",getiparam("screen"),getiparam("orbit"));
    fprintf(ostr,"# \n");
    fprintf(ostr,"# \n");
    fprintf(ostr,"%d %d %d %d %d %d %d\n",ci[0],ci[1],ci[2],ci[3],ci[4],ci[5],ci[6]);
    fprintf(ostr,"# \n");
    fprintf(ostr,"%d\n",getiparam("nstep"));
    fprintf(ostr,"# \n");
    fprintf(ostr,"%d\n",getiparam("dev"));
    fprintf(ostr,"# \n");
    fprintf(ostr,"%d\n",getiparam("sali"));
    strclose(ostr);

    // run!
    
    sprintf(runcmd,"%s",exefile);
    run_sh(runcmd);
}
